File: internal/appdata/appdata.go
package appdata

import (
	"errors"
	"os"
	"path/filepath"
	"strings"
)

// DefaultSupportDir returns the OS-idiomatic config directory for
// Seriesly, without considering any user override:
//   - macOS:   ~/Library/Application Support/Seriesly
//   - Windows: %APPDATA%\Seriesly
//   - Linux:   $XDG_CONFIG_HOME/Seriesly (or ~/.config/Seriesly)
func DefaultSupportDir() (string, error) {
	base, err := os.UserConfigDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(base, "Seriesly"), nil
}

// OverrideFile is the bootstrap redirect file — a single-line text
// file containing the absolute path of the user's chosen config
// directory. Lives inside the platform default so it remains
// findable even after the real config has moved. A user cleaning
// up can delete this file (or its containing dir) and the app will
// fall back to the default on next launch.
func OverrideFile() (string, error) {
	d, err := DefaultSupportDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(d, "config_dir_override"), nil
}

// SupportDir returns the active config directory. Reads the
// override file if present; otherwise falls back to DefaultSupportDir.
func SupportDir() (string, error) {
	override, err := readOverride()
	if err == nil && override != "" {
		return override, nil
	}
	return DefaultSupportDir()
}

// WriteOverride points future launches at dir. Pass "" to clear the
// override and revert to the default on next start. The target
// directory is created if it doesn't exist. Absolute paths only.
func WriteOverride(dir string) error {
	f, err := OverrideFile()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(f), 0o755); err != nil {
		return err
	}
	if dir == "" {
		if rmErr := os.Remove(f); rmErr != nil && !os.IsNotExist(rmErr) {
			return rmErr
		}
		return nil
	}
	if !filepath.IsAbs(dir) {
		return errors.New("config directory path must be absolute")
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	return os.WriteFile(f, []byte(dir+"\n"), 0o644)
}

func readOverride() (string, error) {
	f, err := OverrideFile()
	if err != nil {
		return "", err
	}
	data, err := os.ReadFile(f)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(data)), nil
}

File: src-tauri/src/appdata.rs
//! OS-idiomatic config directory resolution, plus a tiny "override
//! file" indirection that lets users relocate Baudrun's data to an
//! arbitrary path (dotfile-repo, iCloud Drive, Dropbox, etc.).
//!
//!   - macOS:   ~/Library/Application Support/Baudrun
//!   - Windows: %APPDATA%\Baudrun
//!   - Linux:   $XDG_CONFIG_HOME/Baudrun (or ~/.config/Baudrun)
//!
//! The override file (`config_dir_override`) always lives inside the
//! platform default so it remains findable even after the real config
//! has moved. Deleting it reverts to the default on next launch.

use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use thiserror::Error;

const APP_DIR_NAME: &str = "Baudrun";
const OVERRIDE_FILE_NAME: &str = "config_dir_override";

#[derive(Debug, Error)]
pub enum AppDataError {
    #[error("user config directory unavailable on this platform")]
    NoConfigDir,
    #[error("config directory path must be absolute")]
    NotAbsolute,
    #[error(transparent)]
    Io(#[from] io::Error),
}

pub type Result<T> = std::result::Result<T, AppDataError>;

/// Default platform support directory, ignoring any override.
pub fn default_support_dir() -> Result<PathBuf> {
    let base = dirs::config_dir().ok_or(AppDataError::NoConfigDir)?;
    Ok(base.join(APP_DIR_NAME))
}

/// Path of the bootstrap redirect file. A single-line text file
/// containing the absolute path of the user's chosen config
/// directory, or missing when the default is in use.
pub fn override_file() -> Result<PathBuf> {
    Ok(default_support_dir()?.join(OVERRIDE_FILE_NAME))
}

/// Active config directory — honors the override file when present,
/// falls back to [`default_support_dir`] otherwise.
pub fn support_dir() -> Result<PathBuf> {
    match read_override() {
        Ok(Some(path)) => Ok(path),
        _ => default_support_dir(),
    }
}

/// Point future launches at `dir`. Pass `None` to clear the override
/// and revert to the default on next start. Absolute paths only; the
/// target directory is created if it doesn't already exist.
pub fn write_override(dir: Option<&Path>) -> Result<()> {
    let override_path = override_file()?;
    if let Some(parent) = override_path.parent() {
        fs::create_dir_all(parent)?;
    }

    match dir {
        None => match fs::remove_file(&override_path) {
            Ok(()) => Ok(()),
            Err(err) if err.kind() == io::ErrorKind::NotFound => Ok(()),
            Err(err) => Err(err.into()),
        },
        Some(path) => {
            if !path.is_absolute() {
                return Err(AppDataError::NotAbsolute);
            }
            fs::create_dir_all(path)?;
            let mut contents = path.to_string_lossy().into_owned();
            contents.push('\n');
            fs::write(&override_path, contents)?;
            Ok(())
        }
    }
}

fn read_override() -> Result<Option<PathBuf>> {
    let path = override_file()?;
    match fs::read_to_string(&path) {
        Ok(data) => {
            let trimmed = data.trim();
            if trimmed.is_empty() {
                Ok(None)
            } else {
                Ok(Some(PathBuf::from(trimmed)))
            }
        }
        Err(err) if err.kind() == io::ErrorKind::NotFound => Ok(None),
        Err(err) => Err(err.into()),
    }
}
